packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -0,0 +1,342 @@
+        nodesAreCompatible = false;
+        vnodeClassName = className;
+    } else if (!isUndefined(classMap)) {
+        // classMap is used when class is set to static value.
+        const classList = renderer.getClassList(elm);
+        let computedClassName = '';
+
+        // all classes from the vnode should be in the element.classList
+        for (const name in classMap) {
+            computedClassName += ' ' + name;
+            if (!classList.contains(name)) {
+                nodesAreCompatible = false;
+            }
+        }
+
+        vnodeClassName = computedClassName.trim();
+
+        if (classList.length > keys(classMap).length) {
+            nodesAreCompatible = false;
+        }
+    }
+
+    if (!nodesAreCompatible) {
+        logError(
+            `Mismatch hydrating element <${elm.tagName.toLowerCase()}>: attribute "class" has different values, expected "${vnodeClassName}" but found "${
+                elm.className
+            }"`,
+            vnode.owner
+        );
+    }
+
+    return nodesAreCompatible;
+}
+
+function vnodesAndElementHaveCompatibleStyle(vnode: VNode, elm: Element): boolean {
+    const {
+        data: { style, styleDecls },
+        owner: { renderer },
+    } = vnode;
+    const elmStyle = renderer.getAttribute(elm, 'style') || '';
+    let vnodeStyle;
+    let nodesAreCompatible = true;
+
+    if (!isUndefined(style) && style !== elmStyle) {
+        nodesAreCompatible = false;
+        vnodeStyle = style;
+    } else if (!isUndefined(styleDecls)) {
+        const parsedVnodeStyle = parseStyleText(elmStyle);
+        const expectedStyle = [];
+        // styleMap is used when style is set to static value.
+        for (let i = 0, n = styleDecls.length; i < n; i++) {
+            const [prop, value, important] = styleDecls[i];
+            expectedStyle.push(`${prop}: ${value + (important ? ' important!' : '')}`);
+
+            const parsedPropValue = parsedVnodeStyle[prop];
+
+            if (isUndefined(parsedPropValue)) {
+                nodesAreCompatible = false;
+            } else if (!parsedPropValue.startsWith(value)) {
+                nodesAreCompatible = false;
+            } else if (important && !parsedPropValue.endsWith('!important')) {
+                nodesAreCompatible = false;
+            }
+        }
+
+        if (keys(parsedVnodeStyle).length > styleDecls.length) {
+            nodesAreCompatible = false;
+        }
+
+        vnodeStyle = ArrayJoin.call(expectedStyle, ';');
+    }
+
+    if (!nodesAreCompatible) {
+        // style is used when class is bound to an expr.
+        logError(
+            `Mismatch hydrating element <${elm.tagName.toLowerCase()}>: attribute "style" has different values, expected "${vnodeStyle}" but found "${elmStyle}".`,
+            vnode.owner
+        );
+    }
+
+    return nodesAreCompatible;
+}
+
+function throwHydrationError() {
+    assert.fail('Server rendered elements do not match client side generated elements');
+}
```

Comment:
One benefit of having hydration in one place is that we can potentially remove this code for environments that don't use hydration. Svelte has a `hydratable` option that it uses for such optimizations: https://github.com/sveltejs/svelte/pull/6525